### PR TITLE
Support multiple themes

### PR DESCRIPTION
**Pull request by anhvunguyen-preciofishbone**

### CODE REVIEW

**Review by phuctm97 — CHANGES_REQUESTED** [review]
CleanShot 2024-09-28 at 12 14 07

Great initiative! Can you move the Theme Select to the Welcome window, I think it's better that way?

**Comment by anhvunguyen-preciofishbone** [comment]
yeb, will do 🧐

**Comment by phuctm97** [comment]
> yeb, will do 🧐

Thanks! Either at top left or bottom left corner of the window will look great I think!

**Comment by anhvunguyen-preciofishbone** [comment]
image
How about this one? Does this look like what you expected?

**Comment by phuctm97** [comment]
> image How about this one? Does this look like what you expected?

Close! I think it should be on the top right or bottom right corner. Because it's rather a nice to have feature, it shouldn't be on the top left corner.

**Comment by anhvunguyen-preciofishbone** [comment]
> > image How about this one? Does this look like what you expected?
> 
> Close! I think it should be on the top right or bottom right corner. Because it's rather a nice to have feature, it shouldn't be on the top left corner.

Yeb, let me update, move it to top right, thanks for suggestion

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,4 @@
 *.tsbuildinfo
 .env*.local
 .DS_Store
+/package-lock.json
```
Comment (phuctm97) [inline .gitignore]: Please remove this, we only use pnpm in this repository.

File: lib/atom.ts
Comment (phuctm97) [inline lib/atom.ts:1]: We only need to store the theme value, there is no need for theme type

File: lib/provider/ui/header/start-button.tsx [hunk 2]
```diff
@@ -39,6 +39,7 @@ function ConnectedMenuListItem({
   const handleSelect = useCallback(() => {
     openWindow(window);
   }, [window, openWindow]);
+
```
Comment (phuctm97) [inline lib/provider/ui/header/start-button.tsx]: Please revert unnecessary changes

File: lib/provider/ui/index.tsx [hunk 3]
```diff
@@ -43,17 +43,17 @@ const Main = styled.main`
 `;
 
 export function UI({ children }: PropsWithChildren): ReactNode {
+  const [theme] = useAtom(themeAtom);
```
Comment (phuctm97) [inline lib/provider/ui/index.tsx]: Please use `useAtomValue` instead of `useAtom` because we don't need setter in this case.

File: lib/provider/ui/index.tsx [hunk 4]
```diff
@@ -43,17 +43,17 @@ const Main = styled.main`
...
-            <Windows />
-          </Main>
-        </Provider>
+        <Header />
+        <Main>
+          {children}
+          <Windows />
+        </Main>
       </DirectionProvider>
     </ThemeProvider>
```
Comment (phuctm97) [inline lib/provider/ui/index.tsx]: Please keep `Provider`, otherwise we'll have hydration & security issues because all clients share the same `Provider`. You'll want to wrap the provider around this UI component so that useAtom will work.

File: lib/window/react.tsx
Comment (phuctm97) [inline lib/window/react.tsx:1]: Please don't change Window implementation. It may cause other apps to stop working. It's not necessary for this PR, either.

File: lib/provider/ui/windows/welcome.tsx
Comment (phuctm97) [inline lib/provider/ui/windows/welcome.tsx:1]: Please don't change unrelated windows.

File: lib/provider/ui/windows/index.tsx [hunk 5]
```diff
@@ -35,6 +36,9 @@ function Window({ window }: WindowProps): ReactNode {
     case "5xx": {
       return <Exception />;
     }
+    case "Control Panel": {
+      return <ControlPanel />
+    }
```
Comment (phuctm97) [inline lib/provider/ui/windows/index.tsx]: Please put window case before the 404 & 5xx case.

File: lib/provider/ui/windows/control-panel.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                newTheme = themes.default.modernDark;
+                break;
+            }
+            default: {
+                newTheme = themes.default.original;
+            }
+        }
+        setTheme({ type: selectedOption.value, value: newTheme });
+        setSelectedTheme(selectedOption.value);
+    };
```
Comment (phuctm97) [inline lib/provider/ui/windows/control-panel.tsx]: There is no need to use switch cases here, we can just use the keys & values of `await import("react95/dist/themes")` directly.

**Review by phuctm97 — CHANGES_REQUESTED** [review]

File: lib/window/react.tsx
Comment (phuctm97) [inline lib/window/react.tsx:1]: Please do not change Window implementation at all. I don't think we need to change it, right?

File: lib/atom.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import type { Theme } from "react95/dist/types";
+
+import { atom } from "jotai";
+import originalTheme from "react95/dist/themes/original";
```
Comment (phuctm97) [inline lib/atom.ts]: Let's call this `original` instead, so that it's the same as the name exported from `react95/dist/themes`

File: lib/provider/ui/index.tsx [hunk 8]
```diff
@@ -44,17 +44,27 @@ const Main = styled.main`
 
 export function UI({ children }: PropsWithChildren): ReactNode {
   return (
-    <ThemeProvider theme={originalTheme}>
+    <Provider>
+      <InnerUI>
+        {children}
+        <Windows />
```
Comment (phuctm97) [inline lib/provider/ui/index.tsx]: This should have only `{children}`, we should leave the `<Windows />` in the original component.

File: lib/provider/ui/index.tsx [hunk 9]
```diff
@@ -44,17 +44,27 @@ const Main = styled.main`
...
+      </InnerUI>
+    </Provider>
+  );
+}
+
+function InnerUI({ children }: PropsWithChildren): ReactNode {
```
Comment (phuctm97) [inline lib/provider/ui/index.tsx]: Let's call this `Layout` instead.

File: lib/provider/ui/windows/control-panel.tsx [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    return (
+        <StyledWindow window="Control Panel" overflowVisible widthAuto>
+            <GroupBox label="Theme" css={{ zIndex: 100 }}>
+                <Select
+                    options={themeOptions}
+                    onChange={handleThemeChange}
+                    value={selectedTheme}
+                    css="z-index=2"
+                />
+            </GroupBox>
```
Comment (phuctm97) [inline lib/provider/ui/windows/control-panel.tsx]: Please do not use zIndex, overflow, and width auto. These should be automatically controlled already by the layout.

File: lib/provider/ui/windows/control-panel.tsx [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+import { themeAtom } from "~/lib/atom";
+import { Window } from "~/lib/window";
+
+type ThemeType = "original" | "blackAndWhite" | "modernDark";
+
+const themeOptions: SelectOption<ThemeType>[] = [
+    { label: "Default", value: "original" },
+    { label: "Black and white", value: "blackAndWhite" },
+    { label: "Modern Dark", value: "modernDark" }
+];
```
Comment (phuctm97) [inline lib/provider/ui/windows/control-panel.tsx]: We should hard-code these values. These values should be dynamically loaded form `react95/dist/themes`.

File: lib/provider/ui/windows/control-panel.tsx [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+    useEffect(() => {
+        const loadThemes = async (): Promise<void> => {
+            const themes = await import("react95/dist/themes");
+
+            setThemesMap({
+                original: themes.default.original,
+                blackAndWhite: themes.default.blackAndWhite,
+                modernDark: themes.default.modernDark,
+            });
```
Comment (phuctm97) [inline lib/provider/ui/windows/control-panel.tsx]: We should not hard code this, it limits the number of themes we'll have. Instead it should be `setThemes(await imoort("react95/dist/themes")`

**Review by phuctm97 — CHANGES_REQUESTED** [review]

File: lib/atom.ts
Comment (phuctm97) [inline lib/atom.ts:1]: Please rename this to `theme-atom.ts`, the file name should always match the export name.

**Review by phuctm97 — CHANGES_REQUESTED** [review]
Please run `pnpm install` to update the lockfile as well.

**Review by phuctm97 — APPROVED** [review]
